[Win][MiniBrowser] Enable "Download linked file" and "Download Image"

kshukuwa opened the pull request:
#### b98c2d1c85b607d1d7a5836e6948bde2bc9d6294
<pre>
[Win][MiniBrowser] Enable &quot;Download linked file&quot; and &quot;Download Image&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=234887">https://bugs.webkit.org/show_bug.cgi?id=234887</a>

Reviewed by Fujii Hironori.

WinCairo WebKit2 now supports downloading files with 250472@main.
So, we enable &quot;Download linked file&quot; and &quot;Download Image&quot; in MiniBrowser.

* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::invokeDidReceiveResponse):
(WebKit::NetworkDataTaskCurl::setPendingDownloadLocation):
(WebKit::NetworkDataTaskCurl::suggestedFilename const):
(WebKit::NetworkDataTaskCurl::deleteDownloadFile):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.h:
* Tools/MiniBrowser/win/Common.cpp:
(getKnownFolderPath):
* Tools/MiniBrowser/win/Common.h:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
(WebKitBrowserWindow::WebKitBrowserWindow):
(WebKitBrowserWindow::navigationActionDidBecomeDownload):
(WebKitBrowserWindow::navigationResponseDidBecomeDownload):
(WebKitBrowserWindow::contextMenuDidCreateDownload):
(WebKitBrowserWindow::setDownloadClient):
(WebKitBrowserWindow::downloadDecideDestinationWithResponse):
(WebKitBrowserWindow::downloadDidFinish):
(WebKitBrowserWindow::downloadDidFailWithError):
* Tools/MiniBrowser/win/WebKitBrowserWindow.h:

Canonical link: <a href="https://commits.webkit.org/266281@main">https://commits.webkit.org/266281@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed11fed36361044789546fcf0d4842f8c79133bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13397 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13711 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14043 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15134 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/12752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13476 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16219 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13738 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15433 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13564 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14217 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11336 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15822 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11505 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12091 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19141 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12580 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12259 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15472 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12761 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/10621 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12032 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3273 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16355 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12603 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
```diff
@@ -573,6 +578,64 @@ bool WebKitBrowserWindow::canTrustServerCertificate(WKProtectionSpaceRef protect
+    }
+
+    std::wstring folderPath;
+    getKnownFolderPath(FOLDERID_Downloads, folderPath);
```
How should we handle an error of getKnownFolderPath? Can we cancel the download here?

kshukuwa replied:
We can cancel the download by returning nullptr.

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/UIProcess/API/C/WKDownloadRef.cpp#L109-L113

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp#L177

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/NetworkProcess/NetworkProcess.cpp#L2122-L2123

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp#L312-L314

kshukuwa replied:
Fixed

fujii commented on Tools/MiniBrowser/win/Common.cpp:
```diff
@@ -78,6 +78,17 @@ bool getAppDataFolder(_bstr_t& directory)
+{
+    PWSTR path;
+
+    if (FAILED(SHGetKnownFolderPath(id, KF_FLAG_CREATE, 0, &path)))
```
You need to free `path` with `CoTaskMemFree`. See https://learn.microsoft.com/en-us/windows/win32/api/shlobj_core/nf-shlobj_core-shgetknownfolderpath

kshukuwa replied:
Good catch! Fixed.

fujii commented on Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
```diff
@@ -573,6 +578,64 @@ bool WebKitBrowserWindow::canTrustServerCertificate(WKProtectionSpaceRef protect
+    std::wstring folderPath;
+    getKnownFolderPath(FOLDERID_Downloads, folderPath);
+
+    return createWKString(folderPath + L"\\" + filename).leakRef();
```
What will happen if the file already exists? Overwrite without a warning? It seems too dangerous.

kshukuwa replied:
The file is not overwritten because WebKit::AllowOverwrite::No is specified. Instead, it raises a CURLE_WRITE_ERROR error if the file already exists.

https://github.com/WebKit/WebKit/blob/5ec234192a2eebd476418c0dea4a3fc20604b92c/Source/WebKit/UIProcess/API/C/WKDownloadRef.cpp#L114

However, since NetworkDataTaskCurl does not implement overwrite processing, I added setPendingDownloadLocation() and FileSystem::openFile() processing referring to NetworkDataTaskDataURL.

fujii replied:
Does any message boxes appear in this case? What does it say?

kshukuwa replied:
<img width="248" alt="ErrorDialog" src="https://github.com/WebKit/WebKit/assets/113365892/8d18bd8b-8eea-449e-813a-212c02ed76cc">

fujii commented on Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
```diff
@@ -573,6 +578,65 @@ bool WebKitBrowserWindow::canTrustServerCertificate(WKProtectionSpaceRef protect
+
+    std::wstring folderPath;
+    if (!getKnownFolderPath(FOLDERID_Downloads, folderPath))
+        return nullptr;
```
Does any message boxes appear in this case? What does it say?

kshukuwa replied:
No error message is displayed in this case (PolicyAction::Ignore).

fujii replied:
Let's add `MessageBox` here.

kshukuwa replied:
Fixed.

kshukuwa replied:
<img width="215" alt="ErrorDialog-1" src="https://github.com/WebKit/WebKit/assets/113365892/22416ee4-bb80-4ce8-bcc6-9e1ed15834c4">

kshukuwa committed:
[Win][MiniBrowser] Enable "Download linked file" and "Download Image"
https://bugs.webkit.org/show_bug.cgi?id=234887

Reviewed by Fujii Hironori.

WinCairo WebKit2 now supports downloading files with 250472@main.
So, we enable "Download linked file" and "Download Image" in MiniBrowser.

* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::invokeDidReceiveResponse):
(WebKit::NetworkDataTaskCurl::setPendingDownloadLocation):
(WebKit::NetworkDataTaskCurl::suggestedFilename const):
(WebKit::NetworkDataTaskCurl::deleteDownloadFile):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.h:
* Tools/MiniBrowser/win/Common.cpp:
(getKnownFolderPath):
* Tools/MiniBrowser/win/Common.h:
* Tools/MiniBrowser/win/WebKitBrowserWindow.cpp:
(WebKitBrowserWindow::WebKitBrowserWindow):
(WebKitBrowserWindow::navigationActionDidBecomeDownload):
(WebKitBrowserWindow::navigationResponseDidBecomeDownload):
(WebKitBrowserWindow::contextMenuDidCreateDownload):
(WebKitBrowserWindow::setDownloadClient):
(WebKitBrowserWindow::downloadDecideDestinationWithResponse):
(WebKitBrowserWindow::downloadDidFinish):
(WebKitBrowserWindow::downloadDidFailWithError):
* Tools/MiniBrowser/win/WebKitBrowserWindow.h:

Canonical link: https://commits.webkit.org/266281@main
